File: backend/controllers/adminController.js
// backend/controllers/adminController.js
const db = require('../config/db');

// Fetch EVERYTHING for the Admin Dashboard
const getAdminDashboardData = async (req, res) => {
    // Security Check: Only ID 999 (The Admin) is allowed in!
    if (req.user.id !== 999) return res.status(403).json({ message: 'Access Denied: Admins Only' });

    try {
        const [users] = await db.execute('SELECT id, name, email, location, created_at, status FROM users ORDER BY created_at DESC');
        const [shops] = await db.execute('SELECT id, shop_name, email, location, created_at, status FROM shop_owners ORDER BY created_at DESC');
        const [offers] = await db.execute(`
            SELECT offers.id, offers.title, offers.discount_details, shop_owners.shop_name, offers.created_at 
            FROM offers 
            JOIN shop_owners ON offers.shop_id = shop_owners.id 
            ORDER BY offers.created_at DESC
        `);

        res.status(200).json({ users, shops, offers });
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: 'Server error fetching admin data' });
    }
};

// The Ultimate Delete Function
const deleteItem = async (req, res) => {
    if (req.user.id !== 999) return res.status(403).json({ message: 'Access Denied: Admins Only' });
    
    const { type, id } = req.params; // type will be 'user', 'shop', or 'offer'

    try {
        if (type === 'user') {
            await db.execute('DELETE FROM claims WHERE user_id = ?', [id]); // Cascade delete claims
            await db.execute('DELETE FROM users WHERE id = ?', [id]);
        } 
        else if (type === 'shop') {
            await db.execute('DELETE FROM offers WHERE shop_id = ?', [id]); // Cascade delete their offers
            await db.execute('DELETE FROM shop_owners WHERE id = ?', [id]);
        } 
        else if (type === 'offer') {
            await db.execute('DELETE FROM claims WHERE offer_id = ?', [id]); // Cascade delete claims on this offer
            await db.execute('DELETE FROM offers WHERE id = ?', [id]);
        } 
        else {
            return res.status(400).json({ message: 'Invalid deletion type' });
        }

        res.status(200).json({ message: `${type} deleted successfully!` });
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: `Error deleting ${type}.` });
    }
};

// --- APPROVE OR REJECT ACCOUNTS ---
const updateAccountStatus = async (req, res) => {
    if (req.user.id !== 999) return res.status(403).json({ message: 'Access Denied: Admins Only' });
    
    const { type, id } = req.params; // 'user' or 'shop'
    const { status } = req.body;     // 'approved' or 'rejected'

    try {
        const table = type === 'user' ? 'users' : 'shop_owners';
        await db.execute(`UPDATE ${table} SET status = ? WHERE id = ?`, [status, id]);
        res.status(200).json({ message: `${type} has been ${status}!` });
    } catch (error) {
        console.error(error);
        res.status(500).json({ message: 'Error updating status' });
    }
};

module.exports = { getAdminDashboardData, deleteItem, updateAccountStatus };

File: backend-python/controllers/admin_controller.py
# backend-python/controllers/admin_controller.py
from datetime import date, datetime

import aiomysql
from fastapi import HTTPException

from config.db import get_pool


def _require_admin(user: dict) -> None:
    if user.get("id") != 999:
        raise HTTPException(status_code=403, detail="Access Denied: Admins Only")


# ---------------------------------------------------------------------------
# GET ADMIN DASHBOARD DATA
# ---------------------------------------------------------------------------
async def get_admin_dashboard_data(current_user: dict):
    _require_admin(current_user)
    pool = get_pool()
    async with pool.acquire() as conn:
        async with conn.cursor(aiomysql.DictCursor) as cur:
            await cur.execute(
                "SELECT id, name, email, location, created_at FROM users ORDER BY created_at DESC"
            )
            users = await cur.fetchall()

            await cur.execute(
                "SELECT id, shop_name, email, location, created_at, status FROM shop_owners ORDER BY created_at DESC"
            )
            shops = await cur.fetchall()

            await cur.execute(
                """
                SELECT offers.id, offers.title, offers.discount_details,
                       shop_owners.shop_name, offers.created_at
                FROM offers
                JOIN shop_owners ON offers.shop_id = shop_owners.id
                ORDER BY offers.created_at DESC
                """
            )
            offers = await cur.fetchall()

    return {
        "users": _serialize(users),
        "shops": _serialize(shops),
        "offers": _serialize(offers),
    }


# ---------------------------------------------------------------------------
# DELETE AN ITEM (user / shop / offer)
# ---------------------------------------------------------------------------
async def delete_item(item_type: str, item_id: int, current_user: dict):
    _require_admin(current_user)
    pool = get_pool()
    async with pool.acquire() as conn:
        async with conn.cursor() as cur:
            if item_type == "user":
                await cur.execute("DELETE FROM claims WHERE user_id = %s", (item_id,))
                await cur.execute("DELETE FROM users WHERE id = %s", (item_id,))
            elif item_type == "shop":
                await cur.execute("DELETE FROM offers WHERE shop_id = %s", (item_id,))
                await cur.execute("DELETE FROM shop_owners WHERE id = %s", (item_id,))
            elif item_type == "offer":
                await cur.execute("DELETE FROM claims WHERE offer_id = %s", (item_id,))
                await cur.execute("DELETE FROM offers WHERE id = %s", (item_id,))
            else:
                raise HTTPException(status_code=400, detail="Invalid deletion type")

    return {"message": f"{item_type} deleted successfully!"}


# ---------------------------------------------------------------------------
# UPDATE ACCOUNT STATUS (approve / reject)
# ---------------------------------------------------------------------------
async def update_account_status(item_type: str, item_id: int, new_status: str, current_user: dict):
    _require_admin(current_user)
    # Use explicit conditional instead of f-string to prevent SQL injection
    pool = get_pool()
    async with pool.acquire() as conn:
        async with conn.cursor() as cur:
            if item_type == "user":
                await cur.execute("UPDATE users SET status = %s WHERE id = %s", (new_status, item_id))
            else:
                await cur.execute("UPDATE shop_owners SET status = %s WHERE id = %s", (new_status, item_id))
    return {"message": f"{item_type} has been {new_status}!"}


# ---------------------------------------------------------------------------
# Helper
# ---------------------------------------------------------------------------
def _serialize(rows) -> list:
    result = []
    for row in rows:
        serialized = {}
        for k, v in row.items():
            if isinstance(v, (datetime, date)):
                serialized[k] = v.isoformat()
            else:
                serialized[k] = v
        result.append(serialized)
    return result
